aioxmpp/ibr/service.py

```diff
@@ -21,27 +21,123 @@
 
 import asyncio
-
 import aioxmpp
 import logging
-
+from aioxmpp.service import Service
 from . import xso
 
 
 logger = logging.getLogger(__name__)
 
 
-class RegistrationService:
+@asyncio.coroutine
+def get_registration_fields(xmlstream,
+                            timeout=60,
+                            ):
     """
-    Service implementing XMPP In-Band Registration(:xep:`0077`).
+    A query is sent to the server to obtain the fields that need to be
+    filled to register with the server.
+
+    :param xmlstream: Specifies the stream connected to the server where
+                      the account will be created.
+    :type xmlstream: :class:`aioxmpp.protocol.XMLStream`
+
+    :param timeout: Maximum time in seconds to wait for an IQ response, or
+                    :data:`None` to disable the timeout.
+    :type timeout: :class:`~numbers.Real` or :data:`None`
+
+    :return: :attr:`list`
+    """
+
+    iq = aioxmpp.IQ(
+        to=aioxmpp.JID.fromstr(xmlstream._to),
+        type_=aioxmpp.IQType.GET,
+        payload=xso.Query()
+    )
+    iq.autoset_id()
+
+    reply = yield from aioxmpp.protocol.send_and_wait_for(xmlstream,
+                                                          [iq],
+                                                          [aioxmpp.IQ],
+                                                          timeout=timeout)
+    return reply.payload
+
+
+@asyncio.coroutine
+def register(query_xso,
+             xmlstream,
+             timeout=60,
+             ):
+    """
+    Create a new account on the server.
+
+    :param query_xso: XSO with the information needed for the registration.
+    :type query_xso: :class:`xso.Query`
+
+    :param xmlstream: Specifies the stream connected to the server where
+                      the account will be created.
+    :type xmlstream: :class:`aioxmpp.protocol.XMLStream`
+
+    :param timeout: Maximum time in seconds to wait for an IQ response, or
+                    :data:`None` to disable the timeout.
+    :type timeout: :class:`~numbers.Real` or :data:`None`
+    """
+    iq = aioxmpp.IQ(
+        to=aioxmpp.JID.fromstr(xmlstream._to),
+        type_=aioxmpp.IQType.SET,
+        payload=query_xso
+    )
+    iq.autoset_id()
+
+    yield from aioxmpp.protocol.send_and_wait_for(xmlstream,
+                                                  [iq],
+                                                  [aioxmpp.IQ],
+                                                  timeout=timeout)
+
+
+def get_used_fields(payload):
+    """
+    Get a list containing the names of the fields that are used in the
+    xso.Query.
+
+    :param payload: Query object o be
+    :type payload: :class:`xso.Query`
+    :return: :attr:`list`
+    """
+    return [a for a in dir(payload)
+            if isinstance(getattr(payload, a), str) and
+            not a.startswith('__')]
+
+
+def get_query_xso(username, password, aux_fields=None):
```

Comment:
This should be the constructor of Query. Also, I suggest to use ``**kwargs`` instead of ``aux_fields``, but I don’t have a strong opinion on that one.